tiledb/cf/xarray_engine/_writer.py

```diff
@@ -111,31 +128,68 @@ def copy_from_xarray(  # noqa: C901
 
 
 def copy_variable(name, variable, array_wrapper, region):
+    # Check the number of dimensions match.
     if len(array_wrapper.shape) != variable.ndim:
         raise ValueError(
-            f"Cannot write variable '{variable.name}' with {variable.ndim} "
-            f"dimensions to an array with {len(array_wrapper.shape)} dimensions."
+            f"Cannot write variable '{name}' with {variable.ndim} dimensions "
+            f"to an array with {len(array_wrapper.shape)} dimensions."
         )
 
-    # Use the dictionary of region slices to compute the indices of the
-    # full region that is being set in the TileDB array (the target).
-    def get_dimension_slice(dim_name, dim_size):
-        indices = region.get(dim_name, slice(None)).indices(dim_size)
-        return slice(indices[0], indices[1], None)
+    # Get the target region to write the data into.
+    def get_dimension_slice(dim_name, var_dim_size, array_dim_size):
+        """Returns target dimension slice for xarray write.
+
+        Raises a value error if the size of the target slice does not match the
+        size of the source dimension.
+
+        dim_name: Name of the dimension to return the slice for.
+        var_dim_size: Size of the dimension in the source xarray variable.
+        array_dim_size: Size of the domain of the dimension in the target TileDB array.
+        """
+        if var_dim_size > array_dim_size:
+            raise ValueError(
+                f"Cannot write variable '{name}' with shape {variable.shape} on a "
+                f"TileDB array with maximum shape {array_wrapper.shape}."
+            )
+        if dim_name not in region:
+            if var_dim_size != array_dim_size:
+                raise ValueError(
+                    f"Failed to  write variable '{name}' with shape {variable.shape} "
+                    f"on a TileDB array with shape {array_wrapper.shape}. Missing "
+                    f"``region`` value for dimension '{dim_name}' with mismatched "
+                    f"sizes."
+                )
+            return slice(0, array_dim_size, None)
+        else:
+            dim_slice = region[dim_name]
+            start = 0 if dim_slice.start is None else dim_slice.start
+            stop = array_dim_size if dim_slice.stop is None else dim_slice.stop
+            if stop > array_dim_size:
+                raise ValueError(
+                    f"Provided region {dim_slice} for dimension '{dim_name}' is out of "
+                    f"bounds on variable '{name}'. The maximum size of dimension "
+                    f"'{dim_name}' on variable '{name}' is {array_dim_size}."
+                )
```

Comment:
Suggestion: for 1.0, we can build our own error classes so that e.g. consumers have the option of catching a specific `TileDBSliceError` or a more general `ValueError`

Reply:
https://app.shortcut.com/tiledb-inc/story/33108